### ISSUE TITLE
False positive type mismatch on `.flatten()` internals

### ISSUE THREAD

**Issue by HKalbasi** [rust-lang/rust-analyzer#15121]
Repro:
```Rust
pub struct Flatten<I: Iterator<Item: IntoIterator>> {
    inner: FlattenCompat<I, <I::Item as IntoIterator>::IntoIter>,
}

struct FlattenCompat<I, U>;

impl<I, U> Iterator for FlattenCompat<I, U>
where
    I: Iterator<Item: IntoIterator<IntoIter = U, Item = U::Item>>,
    U: Iterator,
{
    type Item = U::Item;
}

impl<I, U> Iterator for Flatten<I>
where
    I: Iterator<Item: IntoIterator<IntoIter = U, Item = U::Item>>,
    U: Iterator,
{
    type Item = U::Item;

    fn next(&mut self) -> Option<U::Item> {
        self.inner.next()
                       // ^^^^ here
    }
}
```
```
expected Option<<<I as Iterator>::Item as IntoIterator>::Item>, found Option<<U as Iterator>::Item>
```

**Comment by HKalbasi** [comment]
Removing the `Item = U::Item` where bound will fix the problem. That is, if we replace the iterator impl with this standalone function
```Rust
fn next<I, U>(this: &mut Flatten<I>) -> Option<U::Item>    
where
    I: Iterator,
    I::Item: IntoIterator<IntoIter = U, Item = U::Item>,
    U: Iterator,

{
    let x = this.inner.next();
    x
}
```
r-a still complains, but after removing the `Item = U::Item` bound
```Rust
fn next<I, U>(this: &mut Flatten<I>) -> Option<U::Item>    
where
    I: Iterator,
    I::Item: IntoIterator<IntoIter = U>,
    U: Iterator,

{
    let x = this.inner.next();
    x
}
```
There will be no type mismatch. Note that the `Item = U::Item` bound is redundant since `IntoIter: Iterator<Item = Self::Item>` is declared in the `IntoIterator` definition.

**Comment by ChayimFriedman2** [comment]
Fixed now.
